libraries/AP_InertialSensor/AP_InertialSensor_tempcal.cpp

```diff
@@ -0,0 +1,491 @@
+    // the calibration process, and the cal_temp, which is the
+    // temperature that the offsets and scale factors was setup for
+    v += polynomial_eval(cal_temp - tmid, coeff);
+}
+
+void AP_InertialSensor::TCal::correct_accel(float temperature, float cal_temp, Vector3f &accel) const
+{
+    correct_sensor(temperature, cal_temp, accel_coeff, accel);
+}
+
+void AP_InertialSensor::TCal::correct_gyro(float temperature, float cal_temp, Vector3f &gyro) const
+{
+    correct_sensor(temperature, cal_temp, gyro_coeff, gyro);
+}
+
+void AP_InertialSensor::TCal::sitl_apply_accel(float temperature, Vector3f &accel) const
+{
+    Vector3f v;
+    correct_sensor(temperature, 0.5*(temp_max+temp_min), accel_coeff, v);
+    accel -= v;
+}
+
+void AP_InertialSensor::TCal::sitl_apply_gyro(float temperature, Vector3f &gyro) const
+{
+    Vector3f v;
+    correct_sensor(temperature, 0.5*(temp_max+temp_min), gyro_coeff, v);
+    gyro -= v;
+}
+
+AP_InertialSensor::TCal::Learn::Learn(TCal &_tcal, float _start_temp) :
+    start_temp(_start_temp),
+    tcal(_tcal)
+{
+    reset(_start_temp);
+}
+
+/*
+  update polyfit with new sample
+ */
+void AP_InertialSensor::TCal::Learn::add_sample(const Vector3f &sample, float temperature, struct LearnState &st)
+{
+    temperature = st.temp_filter.apply(temperature);
+
+    st.sum += sample;
+    st.sum_count++;
+
+    if (st.sum_count < 100 ||
+        temperature - st.last_temp < 0.5) {
+        // wait for more data
+        return;
+    }
+
+    st.sum /= st.sum_count;
+
+    const uint8_t si = &st - &state[0];
+
+    const float T = (temperature + st.last_temp) * 0.5;
+
+    if (si == 0) {
+        // we use the first accel sample as the zero baseline
+        if (accel_start.is_zero()) {
+            accel_start = st.sum;
+            start_temp = T;
+        }
+        st.sum -= accel_start;
+    }
+    
+    const float tmid = 0.5 * (tcal.temp_max + start_temp);
+    const float tdiff = T - tmid;
+
+    AP::logger().Write("TCLR", "TimeUS,I,Si,Temp,TDiff,X,Y,Z",
+                       "s#------",
+                       "F0000000",
+                       "QBBfffff",
+                       AP_HAL::micros64(),
+                       instance(),
+                       si,
+                       T,
+                       tdiff,
+                       st.sum.x, st.sum.y, st.sum.z);
+    
+    
+    st.pfit[0].update(tdiff, st.sum.x);
+    st.pfit[1].update(tdiff, st.sum.y);
+    st.pfit[2].update(tdiff, st.sum.z);
+
+    st.sum.zero();
+    st.sum_count = 0;
+    st.last_temp = temperature;
+
+    if (!is_equal(start_tmax,tcal.temp_max.get())) {
+        // user has changed the TMAX. This will give a bad result for
+        // online learning as the reference temperature (tmid) will
+        // change, so we need to start again
+        reset(T);
+        return;
+    }
+    
+    if (temperature >= tcal.temp_max &&
+        temperature - start_temp >= TEMP_RANGE_MIN) {
```

Comment:
what if user chooses min max temps that don't differ by TEMP_RANGE_MIN? Do we handle and throw the error message for it?

Reply:
it will just wait until it has reached a sufficiently high temperature to meet the min range requirement